data/transactions/application.go

```diff
@@ -0,0 +1,769 @@
+// not valid.
+func (ac *ApplicationCallTxnFields) AddressByIndex(accountIdx uint64, sender basics.Address) (basics.Address, error) {
+	// Index 0 always corresponds to the sender
+	if accountIdx == 0 {
+		return sender, nil
+	}
+
+	// An index > 0 corresponds to an offset into txn.Accounts. Check to
+	// make sure the index is valid.
+	if accountIdx > uint64(len(ac.Accounts)) {
+		err := fmt.Errorf("cannot load account[%d] of %d", accountIdx, len(ac.Accounts))
+		return basics.Address{}, err
+	}
+
+	// accountIdx must be in [1, len(ac.Accounts)]
+	return ac.Accounts[accountIdx-1], nil
+}
+
+// createApplication writes a new AppParams entry and returns application ID
+func (ac *ApplicationCallTxnFields) createApplication(
+	balances Balances, creator basics.Address, txnCounter uint64,
+) (appIdx basics.AppIndex, err error) {
+
+	// Fetch the creator's (sender's) balance record
+	record, err := balances.Get(creator, false)
+	if err != nil {
+		return
+	}
+
+	// Make sure the creator isn't already at the app creation max
+	maxAppsCreated := balances.ConsensusParams().MaxAppsCreated
+	if len(record.AppParams) >= maxAppsCreated {
+		err = fmt.Errorf("cannot create app for %s: max created apps per acct is %d", creator.String(), maxAppsCreated)
+		return
+	}
+
+	// Clone app params, so that we have a copy that is safe to modify
+	record.AppParams = cloneAppParams(record.AppParams)
+
+	// Allocate the new app params (+ 1 to match Assets Idx namespace)
+	appIdx = basics.AppIndex(txnCounter + 1)
+	record.AppParams[appIdx] = basics.AppParams{
+		ApprovalProgram:   ac.ApprovalProgram,
+		ClearStateProgram: ac.ClearStateProgram,
+		LocalStateSchema:  ac.LocalStateSchema,
+		GlobalStateSchema: ac.GlobalStateSchema,
+	}
+
+	// Update the cached TotalStateSchema for this account, used
+	// when computing MinBalance, since the creator has to store
+	// the global state
+	totalSchema := record.TotalAppSchema
+	totalSchema = totalSchema.AddSchema(ac.GlobalStateSchema)
+	record.TotalAppSchema = totalSchema
+
+	// Tell the cow what app we created
+	created := []basics.CreatableLocator{
+		{
+			Creator: creator,
+			Type:    basics.AppCreatable,
+			Index:   basics.CreatableIndex(appIdx),
+		},
+	}
+
+	// Write back to the creator's balance record and continue
+	err = balances.PutWithCreatables(record, created, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	return
+}
+
+func (ac *ApplicationCallTxnFields) applyClearState(
+	balances Balances, sender basics.Address, appIdx basics.AppIndex,
+	ad *ApplyData, steva StateEvaluator,
+) error {
+
+	// Fetch the application parameters, if they exist
+	params, creator, exists, err := getAppParams(balances, appIdx)
+	if err != nil {
+		return err
+	}
+
+	record, err := balances.Get(sender, false)
+	if err != nil {
+		return err
+	}
+
+	// Ensure sender actually has LocalState allocated for this app.
+	// Can't clear out if not currently opted in
+	_, ok := record.AppLocalStates[appIdx]
```

Comment:
don't you need to check if AppLocalStates == nil ?

Reply:
should be OK as is: https://play.golang.org/p/Mw6QlAey-HD

Reply:
> A nil map behaves like an empty map when reading, but attempts to write to a nil map will cause a runtime panic.

This one was kind of eye-opener on go nils vs c++ :)